File: 0110.BalancedBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        '''
        if not root:# or (not root.left and not root.right):
            return True
        if abs(self.depth(root.left)-self.depth(root.right)) > 1:
            return False
        return self.isBalanced(root.left) and self.isBalanced(root.right)
        '''
        self.balance = True
        self.search(root)
        return self.balance

    def depth(self, node):
        if not node:
            return 0
        return max(self.depth(node.left), self.depth(node.right)) + 1

    def search(self, node):
        if not self.balance or not node:
            return 0
        left = self.search(node.left)
        right = self.search(node.right)
        if abs(left-right) > 1:
            self.balance = False
        return max(left, right) + 1

File: 0110.BalancedBinaryTree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
	// Bottom-up, O(n)
    return dfs(root) != -1
}

func dfs(node *TreeNode) int {
    if node == nil {
        return 0
    }
    depthL, depthR := dfs(node.Left), dfs(node.Right)
    if depthL == -1 || depthR == -1 || abs(depthL-depthR) > 1 {
        return -1
    }
    return max(depthL, depthR) + 1
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}